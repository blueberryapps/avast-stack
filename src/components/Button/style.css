.Button {
  color: white;
}

.Button-primary {
  background: blue
}

.Button-secondary {
  background: red;
}

.Button-disabled {
  background: gray
}
