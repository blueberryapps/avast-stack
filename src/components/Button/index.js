import React, { PropTypes as RPT } from 'react';
import './style.css';

const resolveClass = (...classes) => classes.filter(c => !!c).join(' ');

const Button = ({ disabled, kind, children, onClick }) => (
  <button
    className={resolveClass('button', `button-${kind}`, disabled && 'button-disabled')}
    onClick={onClick}
  >
    {children}
  </button>
);

Button.defaultProps = {
  kind: 'primary'
};

Button.propTypes = {
  disabled: RPT.bool,
  children: RPT.node,
  kind: RPT.oneOf(['primary', 'secondary']),
  onClick: RPT.func
};

export default Button;
