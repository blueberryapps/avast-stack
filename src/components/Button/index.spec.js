import { render, shallow } from 'enzyme';
import * as React from 'react';
import Button from './index';

// Snapshot tests
// Looping over test cases is effective and DRY, just respect KISS principle!
// https://en.wikipedia.org/wiki/KISS_principle
describe('<Button /> snapshots', () => {
  const kinds = ['primary', 'secondary'];

  kinds.forEach(kind => (
    it(`should render buttons of kind: ${kind}`, () => {
      expect(render(<Button kind={kind} />)).toMatchSnapshot();
    })
  ));

  const testCases = [
    { kind: 'primary', disabled: true },
    { kind: 'primary', disabled: false },
    { kind: 'secondary', disabled: false },
    { kind: 'secondary', disabled: true }
  ];

  testCases.forEach((props) => {
    it(`should render button with props: ${JSON.stringify(props)}`, () => {
      const wrapper = render(<Button {...props} />);

      // DISCUSS if needed - protects from breaking main functionality?
      expect(wrapper.find('button').length).toBe(1);

      // change breaking functionality can be easily overlooked?
      expect(wrapper).toMatchSnapshot();
    });
  });
});

// Behavior testing
describe('<Button /> behavior', () => {
  it('should register onClick handler', () => {
    const onClickHandler = jest.fn();
    const button = shallow(<Button kind="primary" onClick={onClickHandler} />);

    button.simulate('click');
    expect(onClickHandler.mock.calls.length).toEqual(1);
  });
});
