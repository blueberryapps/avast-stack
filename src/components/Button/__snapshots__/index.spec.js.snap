// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> snapshots should render button with props: {"kind":"primary","disabled":false} 1`] = `
<button
  class="Button Button-primary"
/>
`;

exports[`<Button /> snapshots should render button with props: {"kind":"primary","disabled":true} 1`] = `
<button
  class="Button Button-primary Button-disabled"
/>
`;

exports[`<Button /> snapshots should render button with props: {"kind":"secondary","disabled":false} 1`] = `
<button
  class="Button Button-secondary"
/>
`;

exports[`<Button /> snapshots should render button with props: {"kind":"secondary","disabled":true} 1`] = `
<button
  class="Button Button-secondary Button-disabled"
/>
`;

exports[`<Button /> snapshots should render buttons of these kinds: ["primary","secondary"] 1`] = `
<button
  class="Button Button-primary"
/>
`;

exports[`<Button /> snapshots should render buttons of these kinds: ["primary","secondary"] 2`] = `
<button
  class="Button Button-secondary"
/>
`;
